src/eve/codegen.py

```diff
@@ -315,107 +341,217 @@ def __str__(self) -> str:
+            message += " rendering error."
 
-    def render_template(self, **kwargs: Any) -> str:
-        return self._formatter_.format(self.definition, **kwargs)
+            raise TemplateRenderingError(message, template=self) from e
 
 
-class StringTemplate(Template):
+class StringTemplate(BaseTemplate):
     """Template adapter for `string.Template`."""
 
     definition: string.Template
 
     def __init__(self, definition: Union[str, string.Template], **kwargs: Any) -> None:
+        super().__init__()
         if isinstance(definition, str):
             definition = string.Template(definition)
         assert isinstance(definition, string.Template)
         self.definition = definition
 
-    def render_template(self, **kwargs: Any) -> str:
-        return self.definition.substitute(**kwargs)
+    def render_values(self, **kwargs: Any) -> str:
+        try:
+            return self.definition.substitute(**kwargs)
+        except Exception as e:
+            message = f"<{type(self).__name__}>"
+            if self.definition_loc:
+                message += f" (created at {self.definition_loc[0]}:{self.definition_loc[1]})"
+            try:
+                loc_info = re.search(r"line (\d+), col (\d+)", str(e))
+                message += f" rendering error at template line: {loc_info[1]}, column {loc_info[2]}."  # type: ignore
+            except Exception:
+                message += " rendering error."
 
+            raise TemplateRenderingError(message, template=self) from e
 
-class JinjaTemplate(Template):
+
+class JinjaTemplate(BaseTemplate):
     """Template adapter for `jinja2.Template`."""
 
     definition: jinja2.Template
 
-    def __init__(self, definition: Union[str, jinja2.Template], **kwargs: Any) -> None:
-        if isinstance(definition, str):
-            definition = jinja2.Template(definition)
-        assert isinstance(definition, jinja2.Template)
-        self.definition = definition
-
-    def render_template(self, **kwargs: Any) -> str:
-        return self.definition.render(**kwargs)
-
+    __jinja_env__ = jinja2.Environment(undefined=jinja2.StrictUndefined)
 
-class MakoTemplate(Template):
+    def __init__(self, definition: Union[str, jinja2.Template], **kwargs: Any) -> None:
+        super().__init__()
+        try:
+            if isinstance(definition, str):
+                definition = self.__jinja_env__.from_string(definition)
+            assert isinstance(definition, jinja2.Template)
+            self.definition = definition
+        except Exception as e:
+            message = "Error in JinjaTemplate"
+            if self.definition_loc:
+                message += f" created at {self.definition_loc[0]}:{self.definition_loc[1]}"
+                try:
+                    if hasattr(e, "lineno"):
+                        message += f" (error likely around line: {e.lineno})"  # type: ignore  # assume Jinja exception
```

Comment:
I tried to see this error, but couldn't trigger it.

Reply:
Syntax errors like this one will trigger it:

```python
from eve.codegen import JinjaTemplate

JinjaTemplate("""
asdf
asdf
asdfsadf
{{{}
""")
```